### PR TITLE
task: set security headers

### LINKED ISSUE #107: Set HTTP security headers for Vercel hosted subdomains
To improve the security of the applications, Set HTTP security headers for Vercel hosted subdomains, as none exist currently except HSTS.

Apply Headers for vercel.com hosted sites in next.config.js, as per: https://nextjs.org/docs/advanced-features/security-headers

**Vercel Hosting, security headers to be set for the following Subdomains;**
- https://dapp.klimadao.finance
- https://staging-dapp.klimadao.finance
- https://www.klimadao.finance
- https://staging-site.klimadao.finance

**#Check Current Header Config**
CMD: `curl -i http://dapp.klimadao.finance`

**# Security headers Information,**
**OverviewNFO:** https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers

- Strict-Transport-Security: '_730 days=63072000 s, just this domain_' currently set, 1 year is acceptable for a domain with includeSubDomains, INFO: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Strict-Transport-Security

- X-XSS-Protection: 'None Set', INFO: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-XSS-Protection.
https://github.com/OWASP/CheatSheetSeries/issues/376#issuecomment-602663932
https://github.com/owncloud/core/issues/38236#issuecomment-748451720
https://github.com/github/secure_headers/issues/439

- X-Frame-Options: 'None Set', INFO: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Frame-Options

- X-Content-Type-Options: 'None Set': INFO: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/X-Content-Type-Options

- Referrer-Policy: 'None Set': INFO: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Referrer-Policy

- Content-Security-Policy: 'None Set': Start with a basic CSP policy then harden. INFO: https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP

**Headers: Proposed HTTP security headers;**
{
  key: 'Strict-Transport-Security',
  value: 'max-age=31536000; includeSubDomains'
}
{
  key: 'X-XSS-Protection',
  value: '0'
}
{
  key: 'X-Frame-Options',
  value: 'SAMEORIGIN'
}
{
  key: 'X-Content-Type-Options',
  value: 'nosniff'
}
{
  key: 'Referrer-Policy',
  value: 'strict-origin-when-cross-origin'
}
{
  key: 'Content-Security-Policy',
  value: 'upgrade-insecure-requests'
}

### PR DESCRIPTION
**Pull request by 0xAeterno**
## Description

Configure and explicitly set security headers on our site and dapp.
Share security header config via `@klimadao/lib/config`

Had issues with requiring modules so converted the next config filetype from `.js` to `.mjs` for import statement support, that seemed to work. Spent way too long getting it to work, was stubborn to share config across our apps 😓 

Following security headers have been added:
```js
[
  {
    key: "Referrer-Policy",
    value: "strict-origin-when-cross-origin",
  },
  {
    key: "Strict-Transport-Security",
    value: "max-age=31536000; includeSubDomains",
  },
  {
    key: "X-Content-Type-Options",
    value: "nosniff",
  },
  {
    key: "X-XSS-Protection",
    value: "0",
  },
  {
    key: "Referrer-Policy",
    value: "strict-origin-when-cross-origin",
  },
  {
    key: "Content-Security-Policy",
    value: "upgrade-insecure-requests",
  },
]
```

dapp
```sh
> curl --head http://localhost:3001

HTTP/1.1 200 OK
Referrer-Policy: strict-origin-when-cross-origin
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-XSS-Protection: 0
Content-Security-Policy: upgrade-insecure-requests
Cache-Control: no-store, must-revalidate
X-Powered-By: Next.js
ETag: "adf4-Wb38TORXxNRD8/eBTLnefZMUSPU"
Content-Type: text/html; charset=utf-8
Content-Length: 44532
Vary: Accept-Encoding
Date: Tue, 08 Mar 2022 15:01:08 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

site
```sh
> curl --head http://localhost:3000

HTTP/1.1 200 OK
Referrer-Policy: strict-origin-when-cross-origin
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-XSS-Protection: 0
Content-Security-Policy: upgrade-insecure-requests
Cache-Control: no-store, must-revalidate
X-Powered-By: Next.js
ETag: "17004-Y/QuZtb7zSTC5vlbVR99LEM0fgI"
Content-Type: text/html; charset=utf-8
Content-Length: 94212
Vary: Accept-Encoding
Date: Tue, 08 Mar 2022 15:02:50 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

## Related Ticket

Resolves #107  

## Checklist

- [x] Building the site with `npm run build-site` works without errors
- [x] Building the app with `npm run build-app` works without errors
- [x] I formatted JS and TS files with running `npm run format-all`

### CODE REVIEW

File: app/next.config.mjs [hunk 1]
```diff
@@ -38,15 +42,16 @@ const config = {
...
-    enabled: process.env.ANALYZE === "true",
-  });
-} else {
-  withBundleAnalyzer = (config) => config;
-}
-
-module.exports = withBundleAnalyzer(config);
+export default withBundleAnalyzer({ enabled: process.env.ANALYZE === "true" })(
+  nextConfig
+);
```
Comment (0xAeterno) [inline app/next.config.mjs:57]: @Atmosfearful - ran into a `SyntaxError: Invalid Statement` error trying to keep the previous implementations intention.  I don't think it's too big of a deal to export with the bundle analyser and rely on passing a flag to enable the analyser.

I'll take another stab at it tomr if you're not happy with it, need to sleep now for my meatspace job tomr 😴

**Comment by gord0b** [comment]
'Missing Header
key: 'X-Frame-Options',
value: 'SAMEORIGIN'

Note:
WWW already has a next.config.js in root, has these headers set on source: "/api/block-rate" already,: https://github.com/KlimaDAO/klimadao/blob/staging/site/next.config.js#L21

**Comment by 0xAeterno** [comment]
> 'Missing Header key: 'X-Frame-Options', value: 'SAMEORIGIN'
> 
> Note: WWW already has a next.config.js in root, has these headers set on source: "/api/block-rate" already,: https://github.com/KlimaDAO/klimadao/blob/staging/site/next.config.js#L21

Thanks, I misread what `X-Frame-Options: SAMEORIGIN` implies and added that back in.

Not sure what you mean in your note, could you clarify that please :)
Adding new headers to the `headers` declaration array won't override whats already defined there.

```sh
> curl --head http://localhost:3000/api/block-rate

Access-Control-Allow-Credentials: true
Access-Control-Allow-Origin: *
Access-Control-Allow-Methods: GET
Access-Control-Allow-Headers: X-CSRF-Token, X-Requested-With, Accept, Accept-Version, Content-Length, Content-MD5, Content-Type, Date, X-Api-Version
Referrer-Policy: strict-origin-when-cross-origin
Strict-Transport-Security: max-age=31536000; includeSubDomains
X-Content-Type-Options: nosniff
X-XSS-Protection: 0
Content-Security-Policy: upgrade-insecure-requests
Cache-Control: no-store, must-revalidate
X-Powered-By: Next.js
ETag: "1e6f-oeOqo3KC4s1FbR33NQ1dQ0FwuYg"
Content-Type: text/html; charset=utf-8
Content-Length: 7791
Vary: Accept-Encoding
Date: Tue, 08 Mar 2022 23:30:01 GMT
Connection: keep-alive
Keep-Alive: timeout=5
```

File: lib/config/securityConfig/index.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    key: "Referrer-Policy",
+    value: "strict-origin-when-cross-origin",
+  },
+  {
+    key: "Strict-Transport-Security",
+    value: "max-age=31536000; includeSubDomains",
+  },
+  {
+    key: "Referrer-Policy",
+    value: "strict-origin-when-cross-origin",
```
Comment (Atmosfearful) [inline lib/config/securityConfig/index.ts:12]: duplicate

**Review by Atmosfearful — CHANGES_REQUESTED** [review]
Yesss thank you for taking this.

_controversial opinion:_
I'd just copypasta the headers into both configs and ditch the shared config and `.mjs`
Makes it easier if we want to configure the sites independently (like if we at some point want to allow embedding for `dapp.klimadao.finance/#/offset`)

I'm getting flashbacks to how crazy the nested/inherited configs got in my previous company's monorepo haha

File: site/next.config.mjs [hunk 3]
```diff
@@ -34,6 +43,10 @@ module.exports = {
           },
         ],
       },
+      {
+        source: "/:path*",
```
Comment (Atmosfearful) [inline site/next.config.mjs:47]: Does this override the above headers that I set specifically for /api/block-rate then?

Comment (0xAeterno) [reply site/next.config.mjs:47]: Nope, we're all good here, Header declarations have an additive effect. The function accepts `Headers[]` to allow more configuration granularity between pages.

**Comment by 0xAeterno** [comment]
> Yesss thank you for taking this.
> 
> _controversial opinion:_ I'd just copypasta the headers into both configs and ditch the shared config and `.mjs` Makes it easier if we want to configure the sites independently (like if we at some point want to allow embedding for `dapp.klimadao.finance/#/offset`)
> 
> I'm getting flashbacks to how crazy the nested/inherited configs got in my previous company's monorepo haha

Not controversial at all :)
There's valid points for and against on either approach, but I agree, probably a bit too early to be making the call for shared config. It's easy enough to change over at this stage! Definitely don't want no PTSD like situation either 😂

**Review by Atmosfearful — APPROVED** [review]

**Pull request merged**